pkgs/sdk/server/src/Internal/FDv2DataSources/FDv2ProtocolHandler.cs

```diff
@@ -0,0 +1,354 @@
+        /// </summary>
+        Changeset,
+
+        /// <summary>
+        /// Indicates that an error has been encountered, and it should be logged.
+        /// </summary>
+        Error,
+
+        /// <summary>
+        /// Indicates that the server has expressed an intent to disconnect and the SDK should log the reason.
+        /// </summary>
+        Goodbye,
+
+        /// <summary>
+        /// Indicates that no special action should be taken.
+        /// </summary>
+        None,
+
+        /// <summary>
+        /// Indicates that an internal error has occured and should be logged.  
+        /// </summary>
+        InternalError,
+    }
+
+    internal interface IFDv2ProtocolAction
+    {
+        FDv2ProtocolActionType Action { get; }
+    }
+
+    internal sealed class FDv2ActionChangeset : IFDv2ProtocolAction
+    {
+        public FDv2ProtocolActionType Action => FDv2ProtocolActionType.Changeset;
+        public ChangeSet Changeset { get; }
+
+        public FDv2ActionChangeset(ChangeSet changeset)
+        {
+            Changeset = changeset;
+        }
+    }
+
+    internal sealed class FDv2ActionError : IFDv2ProtocolAction
+    {
+        public FDv2ProtocolActionType Action => FDv2ProtocolActionType.Error;
+
+        public string Id { get; }
+        public string Reason { get; }
+
+        public FDv2ActionError(string id, string reason)
+        {
+            Id = id;
+            Reason = reason;
+        }
+    }
+
+    internal sealed class FDv2ActionGoodbye : IFDv2ProtocolAction
+    {
+        public FDv2ProtocolActionType Action => FDv2ProtocolActionType.Goodbye;
+
+        public string Reason { get; }
+
+        public FDv2ActionGoodbye(string reason)
+        {
+            Reason = reason;
+        }
+    }
+
+    internal sealed class FDv2ActionInternalError : IFDv2ProtocolAction
+    {
+        public FDv2ProtocolActionType Action => FDv2ProtocolActionType.InternalError;
+        public string Message { get; }
+
+        public FDv2ActionInternalError(string message)
+        {
+            Message = message;
+        }
+    }
+
+    internal sealed class FDv2ActionNone : IFDv2ProtocolAction
+    {
+        public FDv2ProtocolActionType Action => FDv2ProtocolActionType.None;
+
+        private FDv2ActionNone()
+        {
+        }
+
+        public static FDv2ActionNone Instance { get; } = new FDv2ActionNone();
+    }
+
+    /// <summary>
+    /// Implements the FDv2 protocol state machine for handling payload communication events.
+    /// See: FDV2PL-payload-communication specification.
+    /// </summary>
+    /// <remarks>
+    /// This handler processes events from Flag Delivery v2 and maintains the protocol state
+    /// as defined in section 4 (Event Handling State Diagram) of the specification.
+    /// The SDK must process events in sequence and only consider payload state valid after
+    /// receiving a payload-transferred event (Requirement 3.3.2).
```

Comment:
I found references to the spec handy, but we might want to consider removing them.